server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -698,57 +722,83 @@ private static ObjectMapper truncateObjectMapper(MapperMergeContext context, Obj
      *
      * @throws IllegalArgumentException if the mapper cannot be flattened
      */
-    List<FieldMapper> asFlattenedFieldMappers(MapperBuilderContext context) {
-        List<FieldMapper> flattenedMappers = new ArrayList<>();
+    List<Mapper> asFlattenedFieldMappers(MapperBuilderContext context, boolean throwOnFlattenableError) {
+        List<Mapper> flattenedMappers = new ArrayList<>();
         ContentPath path = new ContentPath();
-        asFlattenedFieldMappers(context, flattenedMappers, path);
+        asFlattenedFieldMappers(context, flattenedMappers, path, throwOnFlattenableError);
         return flattenedMappers;
     }
 
-    private void asFlattenedFieldMappers(MapperBuilderContext context, List<FieldMapper> flattenedMappers, ContentPath path) {
-        ensureFlattenable(context, path);
+    private static boolean isFlattenable(Optional<Subobjects> subobjects, ObjectMapper mapper, MapperBuilderContext context) {
+        return subobjects.isPresent() && subobjects.get() != Subobjects.ENABLED && mapper instanceof NestedObjectMapper == false;
+    }
+
+    private static boolean throwOnFlattenableError(Optional<Subobjects> subobjects) {
+        return subobjects.isPresent() && subobjects.get() == Subobjects.DISABLED;
+    }
+
+    private void asFlattenedFieldMappers(
+        MapperBuilderContext context,
+        List<Mapper> flattenedMappers,
+        ContentPath path,
+        boolean throwOnFlattenableError
+    ) {
+        var error = checkFlattenable(context);
+        if (error.isPresent()) {
+            if (throwOnFlattenableError) {
+                throw new IllegalArgumentException(
+                    "Object mapper ["
+                        + path.pathAsText(leafName())
+                        + "] was found in a context where subobjects is set to false. "
+                        + "Auto-flattening ["
+                        + path.pathAsText(leafName())
+                        + "] failed because "
+                        + error.get()
+                );
+            }
+            // The object can't be auto-flattened, so it gets added at the current level.
+            // [subobjects=auto] applies auto-flattening to names, so the leaf name may need to change.
+            // Since mapper objects are immutable, we create a clone of the current one with the updated leaf name.
+            flattenedMappers.add(
```

Comment:
Should we attempt to flatten the sub-objects of this object mapper? Even if that object mapper can't be flattened, it may have sub-objects that can be flattened.

Reply:
Since we iterate over all object mappers, this will happen either way. Here,  we don't want to proceed as we'd use the parent object as reference for flattening subobjects. For instance, if "path.to.foo" is auto, we'd want to add "path.to.foo.to.bar" as its sub-field with name "to.bar", instead of adding field "path.to.foo.to.bar" under root directly.

Does this help?

Reply:
This is also covered in tests.